tools/parity_scale_decoder/src/generate_map.rs

```diff
@@ -0,0 +1,386 @@
+//! Exports `generate_map()` function and contains implementation details for it
+
+use std::collections::BTreeSet;
+
+use iroha_core::*;
+use iroha_crypto::*;
+use iroha_data_model::{prelude::*, *};
+use iroha_schema::IntoSchema;
+use iroha_version::*;
+
+use super::*;
+
+/// Trait to retrieve type name
+///
+/// It is used with abusing [inherit impls](https://doc.rust-lang.org/reference/items/implementations.html#inherent-implementations)
+/// to get `None` variant from types, which doesn't implement [`IntoSchema`] and `Some` which does
+trait TypeName {
```

Comment:
this trait seems useless? Can you point me where it's used?

Reply:
No it's need to provide some sort of overloading. It is used here:

```rust
WithTypeName::<$t>::type_name().unwrap_or(stringify!($t).to_owned())
```

This trait will provide `None` for `$t` which does not implement `IntoSchema`

Reply:
I used the idea from this [crate](https://github.com/nvzqz/impls#how-it-works)